### PR TITLE
Added merge capability to structured_data; to handle ingesting partial objects.

### PR DESCRIPTION
**Pull request by dmichaels-harvard**
* Added merge capabilities to structured_data.
* Added Question class to command_utils (factored out of smaht-submitr).
* Refactored out some identifying property related code from portal_object_utils to portal_utils.
* Internalized lookup_strategy related code to structured_data/portal_object_utils/portal_utils.
* Added delete/purge_metadata to portal_utils.Portol, for testing and completeness.

### CODE REVIEW

**Review by willronchetti — APPROVED** [review]
Looks fine but I'd really consider what's trying to be accomplished further downstream with delete/purge metadata

File: dcicutils/portal_utils.py [hunk 1]
```diff
@@ -294,6 +280,20 @@ def post_metadata(self, object_type: str, data: dict, check_only: bool = False)
                                  add_on="check_only=True" if check_only else "")
         return self.post(f"/{object_type}{'?check_only=True' if check_only else ''}", data).json()
 
+    def delete_metadata(self, object_id: str) -> Optional[dict]:
+        if isinstance(object_id, str) and object_id:
+            if self.key:
+                return delete_metadata(obj_id=object_id, key=self.key)
+            else:
+                return self.patch_metadata(object_id, {"status": "deleted"})
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply dcicutils/portal_utils.py:288]: yeah, that was dumb - no need to  check the key - just wrong if not given.

File: dcicutils/portal_utils.py [hunk 2]
```diff
@@ -294,6 +280,20 @@ def post_metadata(self, object_type: str, data: dict, check_only: bool = False)
...
+        return None
+
+    def purge_metadata(self, object_id: str) -> Optional[dict]:
+        if isinstance(object_id, str) and object_id:
+            if self.key:
+                return purge_metadata(obj_id=object_id, key=self.key)
+        return None
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply dcicutils/portal_utils.py:295]: Okay that probably answers my above question - I should probably remove this deletion/purge code from here and when I need it in tests, just implement it there.

File: dcicutils/portal_utils.py [hunk 3]
```diff
@@ -416,6 +419,215 @@ def get_schema_subtype_names(self, type_name: str) -> List[str]:
             return []
         return schemas_super_type_map.get(type_name, [])
 
+    @function_cache(maxsize=100, serialize_key=True)
+    def get_identifying_paths(self, portal_object: dict, portal_type: Optional[Union[str, dict]] = None,
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply dcicutils/portal_utils.py:423]: You mean query the portal for how to query the portal? Just making sure I understand the suggestion ... So like a "identifying_path" API which, for a given type - and object - return the portal path to query for that object ?

**Pull request merged**